### PR TITLE
Fix concurrency issue

### PR DESCRIPTION
**Pull request by zwwhdls**
add lock of each pod. use podLock in csi and manager.

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -41,6 +41,11 @@ var (
 func init() {
 	klog.InitFlags(nil)
 	flag.Parse()
+	podLocks := make([]*sync.Mutex, 1024)
```
Comment (davies) [inline cmd/main.go]: config.PodLocks = make([]sync.Mutex, 1024)

Comment (davies) [reply cmd/main.go]: Then we don't need the following lines.

File: pkg/config/config.go [hunk 2]
```diff
@@ -64,3 +64,11 @@ const (
 	mountPodAnnotationKey  = "juicefs/mount-annotations"
 	mountPodServiceAccount = "juicefs/mount-service-account"
 )
+
+var PodLocks []*sync.Mutex
+
+func GetPodLock(podName string) *sync.Mutex {
+	key := md5.Sum([]byte(podName))
```
Comment (davies) [inline pkg/config/config.go]: key is not a number, you may want to use other Hash function

File: pkg/juicefs/k8sclient/client.go [hunk 3]
```diff
@@ -32,6 +39,10 @@ type K8sClient struct {
 
 func NewClient() (*K8sClient, error) {
 	config, err := rest.InClusterConfig()
+	if config == nil {
+		return nil, status.Error(codes.NotFound, "Can't get kube InClusterConfig")
```
Comment (davies) [inline pkg/juicefs/k8sclient/client.go:46]: this will hide the above error

File: pkg/config/config.go [hunk 4]
```diff
@@ -64,3 +63,12 @@ const (
 	mountPodAnnotationKey  = "juicefs/mount-annotations"
 	mountPodServiceAccount = "juicefs/mount-service-account"
 )
+
+var PodLocks []sync.Mutex
```
Comment (fishincat) [inline pkg/config/config.go]: Why init PodLocks in test file so many times instead of here?

File: pkg/controller/pod_driver.go [hunk 5]
```diff
@@ -315,21 +374,25 @@ func (p *PodDriver) podReadyHandler(ctx context.Context, pod *corev1.Pod) (recon
...
 
-	// recovery for each target
-	for k, target := range pod.Annotations {
-		if k == util.GetReferenceKey(target) {
-			mi := p.mit.resolveTarget(target)
-			if mi == nil {
-				klog.Errorf("pod %s target %s resolve fail", pod.Name, target)
-				continue
-			}
+	_ = doWithinTime(ctx, func() error {
```
Comment (fishincat) [inline pkg/controller/pod_driver.go]: This function can not be executed in concurrence!
Use doWithinTime will case some unexpeced things happen.

File: pkg/juicefs/mount/pod_mount.go [hunk 6]
```diff
@@ -173,64 +183,60 @@ func (p *PodMount) waitUntilMount(jfsSetting *jfsConfig.JfsSetting, volumeId, ta
...
 
+func (p *PodMount) waitUtilPodReady(volumeId string) error {
+	podName := GeneratePodNameByVolumeId(volumeId)
 	// Wait until the mount pod is ready
-	for i := 0; i < 30; i++ {
+	for i := 0; i < 60; i++ {
 		pod, err := p.K8sClient.GetPod(podName, jfsConfig.Namespace)
 		if err != nil {
-			return status.Errorf(codes.Internal, "waitUtilMount: Get pod %v failed: %v", volumeId, err)
+			return status.Errorf(codes.Internal, "waitUtilPodReady: Get pod %v failed: %v", volumeId, err)
```
Comment (davies) [inline pkg/juicefs/mount/pod_mount.go:223]: The pod could be deleted and recovered, so we should retry when error is returned

File: pkg/juicefs/mount/pod_mount.go [hunk 7]
```diff
@@ -173,64 +183,60 @@ func (p *PodMount) waitUntilMount(jfsSetting *jfsConfig.JfsSetting, volumeId, ta
...
+				// pod not exist, create
+				klog.V(5).Infof("createOrAddRef: Need to create pod %s.", podName)
+				newPod := NewMountPod(podName, cmd, mountPath, podResource, config, env, labels, anno, serviceAccount)
+				if newPod.Annotations == nil {
+					newPod.Annotations = make(map[string]string)
+				}
+				newPod.Annotations[key] = target
+				_, err = p.K8sClient.CreatePod(newPod)
+				if err != nil {
+					klog.Errorf("createOrAddRef: Create pod %s err and retry: %v", podName, err)
```
Comment (davies) [inline pkg/juicefs/mount/pod_mount.go:203]: continue?

File: pkg/juicefs/mount/pod_mount.go [hunk 8]
```diff
@@ -173,64 +183,60 @@ func (p *PodMount) waitUntilMount(jfsSetting *jfsConfig.JfsSetting, volumeId, ta
...
+				}
+				return nil
 			}
-			klog.Errorf("get pod err:%v", err)
+			// unexpect error
+			klog.Errorf("createOrAddRef: Get pod %s err and retry: %v", podName, err)
```
Comment (davies) [inline pkg/juicefs/mount/pod_mount.go:208]: continue?

File: pkg/juicefs/mount/pod_mount.go [hunk 9]
```diff
@@ -173,64 +183,60 @@ func (p *PodMount) waitUntilMount(jfsSetting *jfsConfig.JfsSetting, volumeId, ta
...
-		newPod := NewMountPod(podName, cmd, mountPath, podResource, config, env, labels, anno, serviceAccount)
-		if newPod.Annotations == nil {
-			newPod.Annotations = make(map[string]string)
-		}
-		newPod.Annotations[key] = target
-		if _, e := p.K8sClient.CreatePod(newPod); e != nil {
-			return e
-		}
+		// pod exist, add refs
+		return p.AddRefOfMount(target, podName)
```
Comment (davies) [inline pkg/juicefs/mount/pod_mount.go:212]: retry if failed to add ref

File: pkg/juicefs/mount/pod_mount.go [hunk 10]
```diff
@@ -43,7 +43,10 @@ func NewPodMount(client *k8sclient.K8sClient, mounter k8sMount.SafeFormatAndMoun
 }
 
 func (p *PodMount) JMount(jfsSetting *jfsConfig.JfsSetting, volumeId, mountPath string, target string, options []string) error {
-	return p.waitUntilMount(jfsSetting, volumeId, target, mountPath, p.getCommand(jfsSetting, mountPath, options))
+	if err := p.createOrAddRef(jfsSetting, volumeId, target, mountPath, p.getCommand(jfsSetting, mountPath, options)); err != nil {
+		return err
```
Comment (davies) [inline pkg/juicefs/mount/pod_mount.go:47]: Should we call JUmount to cleanup?

File: pkg/juicefs/mount/pod_mount.go (hunk 6 above)
Comment (zwwhdls) [reply pkg/juicefs/mount/pod_mount.go:223]: Call apiServer error is unexpected error, we should return it and wait for kubelet retry

File: pkg/juicefs/mount/pod_mount.go (hunk 7 above)
Comment (zwwhdls) [reply pkg/juicefs/mount/pod_mount.go:203]: ditto

File: pkg/juicefs/mount/pod_mount.go (hunk 8 above)
Comment (zwwhdls) [reply pkg/juicefs/mount/pod_mount.go:208]: ditto

File: pkg/juicefs/mount/pod_mount.go (hunk 9 above)
Comment (zwwhdls) [reply pkg/juicefs/mount/pod_mount.go:212]: ditto

**Pull request merged**